src/wallet/rpcnames.cpp

```diff
@@ -483,11 +523,81 @@ name_firstupdate ()
   if (name.size () > MAX_NAME_LENGTH)
     throw JSONRPCError (RPC_INVALID_PARAMETER, "the name is too long");
 
-  const valtype rand = ParseHexV (request.params[1], "rand");
-  if (rand.size () > 20)
-    throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+  const bool fixedRand = request.params.size () >= 2 && !request.params[1].isNull ();
+  const bool fixedTxid = request.params.size () >= 3 && !request.params[2].isNull ();
+  valtype rand(20);
+  if (fixedRand)
+    {
+      rand = ParseHexV (request.params[1], "rand");
+      if (rand.size () > 20)
+        throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+    }
+
+  uint256 prevTxid = uint256::ZERO; // if it can't find a txid, force an error
+  if (fixedTxid)
+    {
+      prevTxid = ParseHashV (request.params[2], "txid");
+    }
+  else
+    {
+      // Code slightly duplicates name_scan, but not enough to be able to refactor.
+      /* Make sure the results are valid at least up to the most recent block
+         the user could have gotten from another RPC command prior to now.  */
+      pwallet->BlockUntilSyncedToCurrentChain ();
+
+      LOCK (cs_main);
+
+      for (const auto& item : pwallet->mapWallet)
+        {
+          const CWalletTx& tx = item.second;
+          if (!tx.tx->IsNamecoin ())
+            continue;
+
+          CScript output;
+          CNameScript nameOp;
+          int nOut = -1;
+          for (unsigned i = 0; i < tx.tx->vout.size (); ++i)
+            {
+              CScript curOutput = tx.tx->vout[i].scriptPubKey;
+              const CNameScript cur(curOutput);
+              if (!cur.isNameOp ())
+                continue;
+              if (cur.getNameOp () != OP_NAME_NEW)
+                continue;
+              if (nOut != -1) {
+                LogPrintf ("ERROR: wallet contains tx with multiple"
+                           " name outputs");
+                continue;
+              }
+              nameOp = cur;
+              nOut = i;
+              output = curOutput;
+            }
+
+          if (nOut == -1)
+            continue; // no name outputs found
 
-  const uint256 prevTxid = ParseHashV (request.params[2], "txid");
+          const int depth = tx.GetDepthInMainChain ();
+          if (depth <= 0)
+            continue; // not confirmed
+
+          if (!fixedRand)
+            assert (!getNameSalt (pwallet, name, output, rand)); // we should have the private key - we're sending from it!
```

Comment:
Two issues here:
1. `assert` should not be used for things with side effects, as it can be compiled out entirely with `NDEBUG`.  You can use `Assert` from `util/check.h` instead I think, or just refactor the code explicitly here to first call `getNameSalt` and then `assert` the result.
2. What happens if someone `name_new`'s to someone else's wallet?  Would that trigger a crash for the receiver (which we obviously don't want, as it could even be a DoS vector)?

Reply:
1) `assert()` can't actually be compiled out, since bitcoin doesn't compile with NDEBUG:
```
// Assumption: We assume that the macro NDEBUG is not defined.                                                                                                                
// Example(s): We use assert(...) extensively with the assumption of it never
//             being a noop at runtime.
#if defined(NDEBUG)
# error "Bitcoin cannot be compiled without assertions."
#endif
```
~compat/assumptions.h

2) No, since this is in name_firstupdate. We will always have the key to spend the input to the transaction, or it won't be solvable.

Either they're sending it to an address in your wallet, in which case it's OK, or they're sending it to an address not in your wallet, in which case it won't even be found here.

Reply:
1. Yes that's true, but still I believe it is bad style, as C programmers reading the code may not know about this special case.  And it may be removed at some point.  IMHO, `assert` should simply never be used with side effects at all, as a general matter of style.  Especially if there is a simple alternative, `Assert`, available here.

2. Not sure I understand your argument.  The salt will be derived from the private key of the `name_new` output, right?  But that is not guaranteed to be in your wallet (probably my wording was confusing previously).  For instance, it could be that the `name_new` output is to an address not in your wallet, but there is an additional NMC output to some of your addresses.  In that case, the transaction will be part of your wallet history and will be found, but you won't be able to generate the salt.  Or it could be a watch-only address you imported.  But in any case, I think it would be best to simply soft-fail (by just ignoring this transaction and ultimately not finding a suitable rand) instead of `assert` in this case in general.

Reply:
1) It is a bit gross I guess. Even cleaner to just copy it to a new variable and then assert on it, so I'll do that instead.

2) Oh, yeah, that is possible. I didn't think of that. Yes, it could just `continue` if it doesn't have the privkey. Since otherwise, any messed up name output that somehow ended up in your wallet will spoil the whole bunch.